cmd/dockerd/hack/malformed_host_override.go

```diff
@@ -15,12 +15,79 @@ type MalformedHostHeaderOverride struct {
 // connection and keeps track of the first read from http.Server
 // which just reads the headers.
 type MalformedHostHeaderOverrideConn struct {
-	net.Conn
+	net.UnixConn
 	first bool
 }
 
 var closeConnHeader = []byte("\r\nConnection: close\r")
 
+// checkOldHost do check whether an old client sends an invalid host head or not
+func checkOldHost(c int, b []byte) (n int, err error) {
+	var (
+		start, end    int
+		firstLineFeed = -1
+		buf           []byte
+	)
+	for i := 0; i <= c-1-7; i++ {
+		if b[i] == '\n' && firstLineFeed == -1 {
+			firstLineFeed = i
+		}
+		if b[i] != '\n' {
+			continue
+		}
+
+		if b[i+1] == '\r' && b[i+2] == '\n' {
+			return c, nil
+		}
+
+		if b[i+1] != 'H' {
+			continue
+		}
+		if b[i+2] != 'o' {
+			continue
+		}
+		if b[i+3] != 's' {
+			continue
+		}
+		if b[i+4] != 't' {
+			continue
+		}
+		if b[i+5] != ':' {
+			continue
+		}
+		if b[i+6] != ' ' {
+			continue
+		}
+		if b[i+7] != '/' {
+			continue
+		}
+		// ensure clients other than the docker clients do not get this hack
+		if i != firstLineFeed {
+			return c, nil
+		}
+		start = i + 7
+		// now find where the value ends
+		for ii, bbb := range b[start:c] {
+			if bbb == '\n' {
+				end = start + ii
+				break
+			}
+		}
+		buf = make([]byte, 0, c+len(closeConnHeader)-(end-start))
+		// strip the value of the host header and
+		// inject `Connection: close` to ensure we don't reuse this connection
+		buf = append(buf, b[:start]...)
+		buf = append(buf, closeConnHeader...)
+		buf = append(buf, b[end:c]...)
+		copy(b, buf)
+		break
+	}
+	if len(buf) == 0 {
+		return c, nil
+	}
+	return len(buf), nil
```

Comment:
Seems weird to allocate buffers and only use them to get the length.

Reply:
`buf` is used in here `copy(b, buf)`, but I also change to a more reasonable way.

Reply:
Ah your right, thinking of append, but copy is ok.
Can we update the function name? To me a check shouldn't be modifying things and makes it confusing... maybe `stripInvalidHeader`

Reply:
>  maybe stripInvalidHeader

OK